File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        current = head
        sz = 1
        while current.next != None: 
            current = current.next
            sz += 1

        if sz == 1:
            return None
        elif sz == n:
            return head.next
            
        target = sz
        current = head
        while target-1 > n:
            current = current.next
            target -= 1

        removed = current.next
        current.next = removed.next

        return head

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null){
            return null;
        }
        ListNode curr = head;
        ListNode prev;
        int count = 0;
        while (curr != null){
            count ++;
            prev = curr;
            curr = curr.next;
        }
        
        ListNode tempHead = new ListNode(0, head);
        curr = head;
        prev = tempHead;
        int target = count - n;
        count = 0;
        while (curr != null){
            if (count == target){
                prev.next = curr.next;
                break;
            }
            count++;
            prev = curr;
            curr = curr.next;
        }
        return tempHead.next;
        
    }
}